Перевод строки в Чтении/Записи текста

dmpas opened the pull request.

dmpas committed:
В ЗаписьТекста доработан учёт разделителей строк.

dmpas committed:
Добавлен тест для переводов строк в чтении и записи текста

dmpas committed:
ЧтениеТекста: добавлена обработка конца строк

dmpas committed:
Доработан конструктор для ЧтениеТекста

dmpas committed:
Исправлен тест `text-document`.

Тест исправлен в соответствии с новым поведением объекта ЧтениеТекста.

dmpas committed:
Добавлен тест для ЧтениеТекста

dmpas committed:
Упрощён алгоритм обработки конца строк

dmpas committed:
Исправлен конец строк по-умолчанию в Текстовом Документе.

В Текстовом документе по-умолчанию подставляется системный перевод строк
вместо ранее зашитого ВК+ПС.

dmpas committed:
Расширен набор тестов для Записи Текста

dmpas committed:
Расширен набор тестов для ЧтениеТекста

dmpas committed:
Для ЧтениеТекста добавлен тест на совместимость.

Тест на разное поведение конструкторов ЧтениеТекста, направленный на
совместимость поведения с Платформой 1С.

dmpas committed:
Добавлено экранирование в тесте text-read.os

dmpas committed:
Доработки по Чтению Текста

dmpas committed:
ЧтениеТекста и ЗаписьТекста сделаны публичными

dmpas committed:
Исправлена опечатка в Текстовом Документе

dmpas committed:
Тесты приведены к единообразию.

Для простоты понимания тестов для всех систем используется единый
перевод строк ВК+ПС.

dmpas committed:
Перевод строк по-умолчанию установлен в ВК+ПС.

Во всех средах по-умолчанию используется перевод строк ВК+ПС для
совместимости и единообразия.

dmpas committed:
Исправлено падение теста CGI

dmpas committed:
Перенесён тест ПрочитатьИзФайлаСРазнымиПереводамиСтрок.

Тест ПрочитатьИзФайлаСРазнымиПереводамиСтрок перенесён из xmlwrite в
text-write.

dmpas commented:
Важное замечание: по-умолчанию для чтения и записи текста под ВСЕМИ осями будет ВК+ПС. Так что скрипты под linux и osx (есть такие вообще?!) надо будет перепроверить на работоспособность.

pumbaEO commented:
Т.е. если будет только ПС в файле, без ВК, то неправильно строки определит?

dmpas commented:
@pumbaEO в общем случае, всё будет в порядке - просто ПС скушается без вопросов. Но всегда есть нюансы, когда начинаешь играться парой разделителей в записи и парой разделителей в чтении.

artbear commented:
Пока тесты не смотрел.

> В: Т.е. если будет только ПС в файле, без ВК, то неправильно строки определит?
> О: в общем случае, всё будет в порядке - просто ПС скушается без вопросов

Вопрос: есть ли тест на этот сценарий?

dmpas commented:
@artbear нет, именно такого нет. сейчас добавлю.

dmpas committed:
Добавлен тест на правильное чтение по-умолчанию.

Тест проверяет, что ЧтениеТекста с разделителем по-умолчанию правильно
считывает текст с переводами строки только ПС.

EvilBeaver commented on src/ScriptEngine.HostedScript/Library/TextReadImpl.cs:
```diff
@@ -65,26 +100,30 @@ private void RequireOpen()
             var reader = new TextReadImpl();
-            reader.Open(path.AsString(), null);
+            if (lineDelimiter != null)
+                reader.AnalyzeDefaultLineFeed = false;
```
Это свойство только для оптимизации CustomLineFeedReader-а?

dmpas replied:
Это свойство для совместимости поведения с 1С. Она в одном случае отрабатывает самостоятельно все переводы строк, в других - нет.

Тут можно глянуть: ТестДолжен_ПроверитьСовместимостьПоведения в text-read

EvilBeaver commented on src/ScriptEngine.HostedScript/Library/TextWriteImpl.cs:
```diff
@@ -72,20 +78,29 @@ public void Close()
+            Write (what);
+
+            var sDelimiter = _lineDelimiter;
+            if (delimiter != null && delimiter.GetRawValue ().DataType != DataType.Undefined)
```
Здесь GetRawValue не нужен. Он нужен только если хотим проверить реальный тип через **is** или **as**
